# Topic Modeling and Natural Language Processing with Twitter Data

##  Jason Anastasopoulos
##  April 2nd 2018
### Email: ljanastas@princeton.edu

The code below provides a brief introduction on acquiring Twitter data using the twitter API via Python. For this exercise I will be acquiring Donald Trump's tweets and will try to figure out what the topics his tweets are using the Latent Dirichlet Allocation  Topic Model.

In [3]:
import csv
import os,re,csv
from nltk.tokenize import RegexpTokenizer
from stop_words import get_stop_words
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from gensim import corpora, models
import gensim
import numpy as np
import scipy
import math
import matplotlib.pyplot as plt
import plotly.plotly as py
import twitter
import json

Here we enter our Twitter credentials. These can be acquired through 

In [4]:
api = twitter.Api(consumer_key='mkz0izzVKDRzkrR4GoyN9FStT',
                      consumer_secret='4A1YGFEixYmyUNf2idYC33GKCuFoyJkyKpQVXIXCpDedZe0nOt',
                      access_token_key='18249358-xZGyGz8sWmQ9oJ1TBsLKEczwtO24aJ0Q4waDbjxAd',
                      access_token_secret='uqH7cC5BLS65iuAEPEv4TXEtUZvFD80wH03xkqiB7SP7Y')

print(api.VerifyCredentials())

{"created_at": "Fri Dec 19 19:38:51 +0000 2008", "description": "Microsoft Visiting Professor @PrincetonCITP\u25aa\ufe0fAssistant Professor @UGA_PA_Policy & Political Science \u25aa\ufe0f political economy\u25aa\ufe0f machine learning\u25aa\ufe0f causal inference", "favourites_count": 798, "followers_count": 486, "friends_count": 541, "geo_enabled": true, "id": 18249358, "id_str": "18249358", "lang": "en", "listed_count": 33, "location": "Princeton, NJ", "name": "Jason Anastasopoulos", "profile_background_color": "C0DEED", "profile_background_image_url": "http://pbs.twimg.com/profile_background_images/561147677/sproul.jpg", "profile_background_image_url_https": "https://pbs.twimg.com/profile_background_images/561147677/sproul.jpg", "profile_banner_url": "https://pbs.twimg.com/profile_banners/18249358/1510588727", "profile_image_url": "http://pbs.twimg.com/profile_images/953308288044683264/pfRYpPeN_normal.jpg", "profile_image_url_https": "https://pbs.twimg.com/profile_images/95330828804

Search the Twitter API using keywords.

In [5]:
search = api.GetSearch("Oscars") # Replace happy with your search
for tweet in search:
    print(tweet.id, tweet.text)
    
len(search)

(976316028010426368, u"This is the theater where the Oscars were held. No offense to Hollywood's best &amp; brightest, but I think we had a be\u2026 https://t.co/ZhUjHWbeQH")
(976324353821429762, u'Carol Kane at the 1976 Oscars is a real mood https://t.co/oF4rzToGUE')
(975918392136802304, u'Would @TiffanyHaddish be the next host for the #Oscars? \n\nShe shares her thoughts at the #TwitterHouse at #SXSW.\u2026 https://t.co/KxkCMtUcvh')
(976616349442199552, u'RT @TwitterTV: Would @TiffanyHaddish be the next host for the #Oscars? \n\nShe shares her thoughts at the #TwitterHouse at #SXSW. #PayTheLady\u2026')
(976616345956683781, u'RT @rilaws: Carol Kane at the 1976 Oscars is a real mood https://t.co/oF4rzToGUE')
(976616310057656320, u'RT @GMA: WAKANDA FOREVER! @chadwickboseman brings a bit of @theblackpanther to the #Oscars red carpet! https://t.co/6YJWYVSrEb https://t.co\u2026')
(976616273940566017, u'RT @VinceSchilling: Friday on Native Trailblazers Radio\n\n"The Real Story Behind the #O

15

The Python twitter library has a lot of cool functions that you can use and learn about through the help() function

In [18]:
help(api.GetUserTimeline)


Help on method GetUserTimeline in module twitter.api:

GetUserTimeline(self, user_id=None, screen_name=None, since_id=None, max_id=None, count=None, include_rts=True, trim_user=False, exclude_replies=False) method of twitter.api.Api instance
    Fetch the sequence of public Status messages for a single user.
    
    The twitter.Api instance must be authenticated if the user is private.
    
    Args:
      user_id (int, optional):
        Specifies the ID of the user for whom to return the
        user_timeline. Helpful for disambiguating when a valid user ID
        is also a valid screen name.
      screen_name (str, optional):
        Specifies the screen name of the user for whom to return the
        user_timeline. Helpful for disambiguating when a valid screen
        name is also a user ID.
      since_id (int, optional):
        Returns results with an ID greater than (that is, more recent
        than) the specified ID. There are limits to the number of
        Tweets which c

In [10]:
t = api.GetUserTimeline(screen_name="realDonaldTrump", count=5000)
tweets = [i.AsDict() for i in t]

tweettext = [i["text"] for i in tweets]

tweettext[2:4]

[u'\u201cSpecial Council is told to find crimes, whether a crime exists or not. I was opposed to the selection of Mueller t\u2026 https://t.co/KYsOPqiVBG',
 u'...there was no probable cause for believing that there was any crime, collusion or otherwise, or obstruction of ju\u2026 https://t.co/zZFkhcS0jf']

Loads the text data which are all referenda in New Mexico from 2000-2014. Splits referenda using "####" demarcation, extracts each bill and places it into a list.

## Cleaning the text

In [11]:
######## So far so good not lets clean this up ###
tokenizer = RegexpTokenizer(r'\w+')

# create English stop words list
en_stop = get_stop_words('en')

# Create p_stemmer of class PorterStemmer
p_stemmer = PorterStemmer()
texts = []

Here we are pre-processing the text by creating a tokenizer that splits the documents up into tokens (words or phrases), creating a dictionary of stop words and creating a "stemmer" which stems the words (ie removing "-ing" endings etc. We also remove extraneous "bill related" words such as "propXX_XXXX".

In [12]:
for i in tweettext:
    #print "Processing",i
    # clean and tokenize document string
    tokens = tokenizer.tokenize(i)
    # remove all numbers
    tokens = [x for x in tokens if not (x.isdigit() or x[0] == '-' and x[1:].isdigit())]
    # remove structural words
    tokens = [x for x in tokens if len(x) > 1]
    tokens = [x.lower() for x in tokens]
    tokens = [x for x in tokens if 'http' not in x]
    tokens = [x for x in tokens if x not in "_"]
    # remove stop words from tokens
    stopped_tokens = [i for i in tokens if not i in en_stop]
    # stem tokens
    stemmed_tokens = [p_stemmer.stem(i) for i in stopped_tokens]
    # add tokens to list
    texts.append(stemmed_tokens)

dictionaryall = corpora.Dictionary(texts)

corpusall = [dictionaryall.doc2bow(text) for text in texts]

texts[1]



[u'can',
 u'help',
 u'solv',
 u'problem',
 u'north',
 u'korea',
 u'syria',
 u'ukrain',
 u'isi',
 u'iran',
 u'even',
 u'come',
 u'arm',
 u'race',
 u'bush',
 u'co',
 u'hrz6vrjjvc']

In [32]:
texts[4]

[u'depart',
 u'justic',
 u'urg',
 u'suprem',
 u'court',
 u'least',
 u'hear',
 u'driver',
 u'licens',
 u'case',
 u'illeg',
 u'immi',
 u'co',
 u'teayq9in3t']

In [13]:
tweettext[0]

u'.....They can help solve problems with North Korea, Syria, Ukraine, ISIS, Iran and even the coming Arms Race. Bush\u2026 https://t.co/hrZ6vrJjVC'

This code performs tokenization, stop word removal and number removal and places the corpora into a clean list that will be ready for analysis using the Latent Dirichlet Allocation. 

Notice that there are two sets of texts that are jointly modeled "commercetexts" which are the "business friendly" Chamber of Commerce bills and "nmtexts" which are the New Mexico state referenda.

## Estimating the Latent Dirichlet Allocation model

In [28]:
# generate LDA model
ldamodelall = gensim.models.ldamodel.LdaModel(corpusall, num_topics=7, id2word = dictionaryall, passes=20,
                                              minimum_probability=0)



The code above estimates a 5 topic topic model using Trump's tweets

In [29]:
# 5 topics from Trump's tweets derived from the top 5 words

print(ldamodelall.print_topics(num_topics=5, num_words=5))

[(6, u'0.041*"co" + 0.018*"great" + 0.012*"countri" + 0.012*"will" + 0.009*"meet"'), (1, u'0.038*"co" + 0.008*"better" + 0.008*"collus" + 0.008*"today" + 0.006*"much"'), (5, u'0.060*"co" + 0.018*"job" + 0.012*"trade" + 0.011*"thank" + 0.008*"year"'), (4, u'0.043*"co" + 0.015*"great" + 0.011*"north" + 0.011*"korea" + 0.009*"just"'), (2, u'0.050*"co" + 0.015*"will" + 0.011*"state" + 0.011*"foxnew" + 0.009*"just"')]


Prints the first 5 topics from the full model.

- Topic 1: "great" + 0.012*"countri" + 0.012*"will" + 0.009*"meet"' - Diplomatic Relations
- Topic 2: "better" + 0.008*"collus" + 0.008*"today" + 0.006*"much"' -  Russia Special Investigation
- Topic 3: "job" + 0.012*"trade" + 0.011*"thank" + 0.008*"year"' - Economy
- Topic 4: 0.015*"great" + 0.011*"north" + 0.011*"korea" + 0.009*"just"' - North Korea
- Topic 5: 0.015*"will" + 0.011*"state" + 0.011*"foxnew" + 0.009*"just"' - Media

Plot the topic distribution for each tweet. 